### PR TITLE
[#172331475] Barcode generation on Receipt Prescription Data

### PR DESCRIPTION
**Pull request by AleDore**
Added utility for generating svg/png barcode images server side, in order to add image data to messages containing Receipt Prescription Data.

### CODE REVIEW

File: src/utils/barcode.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  const document = new DOMImplementation().createDocument(
+    "http://www.w3.org/1999/xhtml",
+    "html",
+    null
+  );
+  const svgNode = document.createElementNS("http://www.w3.org/2000/svg", "svg");
+
+  JsBarcode(svgNode, text, {
+    xmlDocument: document
+    // tslint:disable-next-line: no-any
```
Comment (gunzip) [inline src/utils/barcode.ts:26]: can you add a comment with a link to the relevant issue on wrong typings ?

File: src/utils/barcode.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
+import { createCanvas } from "canvas";
+import * as t from "io-ts";
+import * as JsBarcode from "jsbarcode";
+import { DOMImplementation, XMLSerializer } from "xmldom";
+
+export const BarcodeOutput = t.interface({
```
Comment (gunzip) [inline src/utils/barcode.ts:6]: don't create a io-ts type if you don't need decode, just use a plain typescript interface

File: src/utils/barcode.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  png: t.string,
+  pngMimeType: t.string,
+  svg: t.string,
+  svgMimeType: t.string
+});
+
+export type BarcodeOutput = t.TypeOf<typeof BarcodeOutput>;
+
+export function toBarcode(text: string): BarcodeOutput {
```
Comment (gunzip) [inline src/utils/barcode.ts:15]: can this method throw exceptions ? can you make it return an Either in case ? (tryCatch2v)

File: src/utils/barcode.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
+import { createCanvas } from "canvas";
+import * as t from "io-ts";
+import * as JsBarcode from "jsbarcode";
+import { DOMImplementation, XMLSerializer } from "xmldom";
+
+export const BarcodeOutput = t.interface({
+  png: t.string,
+  pngMimeType: t.string,
```
Comment (gunzip) [inline src/utils/barcode.ts:8]: these are t.literal() not t.string

File: src/utils/barcode.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  const svg = svgBuffer.toString("base64");
+  const canvas = createCanvas(200, 200);
+  JsBarcode(canvas, text);
+  const png = canvas.toBuffer().toString("base64");
+  return {
+    png,
+    pngMimeType: "image/png",
+    svg,
+    svgMimeType: "image/svg+xml"
+  } as BarcodeOutput;
```
Comment (gunzip) [inline src/utils/barcode.ts:40]:
```suggestion
  };
```

File: src/utils/barcode.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+  svgMimeType: string;
+}
+
+interface IBwipOptions {
+  bcid: string;
+  text: string;
+  includetext: boolean;
+}
+
+const toBase64Png = (options: IBwipOptions) => {
```
Comment (cloudify) [inline src/utils/barcode.ts:19]: you can omit the first curly brace + `return`:

```ts
const toBase64Png = (options: IBwipOptions) => tryCatch(...)
```

File: src/utils/barcode.ts [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+            return reject(new Error("Cannot generate png barcode"));
+          }
+        })
+      );
+    },
+    errs => new Error(`Cannot generate png barcode|${errs}`)
+  );
+};
+
+const toBase64Svg = (text: string) => {
```
Comment (cloudify) [inline src/utils/barcode.ts:36]: same here

File: src/utils/barcode.ts [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+interface IBwipOptions {
+  bcid: string;
+  text: string;
+  includetext: boolean;
+}
+
+const toBase64Png = (options: IBwipOptions) => {
+  return tryCatch(
+    () => {
```
Comment (cloudify) [inline src/utils/barcode.ts:21]: same here

```ts
() => new Promise...
```

File: src/utils/barcode.ts [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+      return new Promise<string>((resolve, reject) =>
+        bwipjs.toBuffer(options, (err, png) => {
+          if (err == null) {
+            return resolve(png.toString("base64"));
+          } else {
+            return reject(new Error("Cannot generate png barcode"));
```
Comment (cloudify) [inline src/utils/barcode.ts:27]: probably better to pass `err` along instead of swallowing it?

File: src/utils/barcode.ts [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+  bcid: string;
+  text: string;
+  includetext: boolean;
+}
+
+const toBase64Png = (options: IBwipOptions) => {
+  return tryCatch(
+    () => {
+      return new Promise<string>((resolve, reject) =>
+        bwipjs.toBuffer(options, (err, png) => {
```
Comment (cloudify) [inline src/utils/barcode.ts:23]: I believe you can use `util.promisify` to convert the callback to a promise

File: package.json [hunk 11]
```diff
@@ -66,6 +66,7 @@
     "io-ts-types": "^0.4.7",
     "italia-ts-commons": "^6.3.1",
     "json-set-map": "^1.0.2",
+    "jsbarcode": "3.11.0",
```
Comment (gunzip) [inline package.json:69]: do we need this ?

File: package.json [hunk 12]
```diff
@@ -79,11 +80,14 @@
     "ulid": "^2.2.2",
     "validator": "^10.8.0",
     "winston": "^3.0.0",
-    "xmldom": "^0.1.27"
+    "xmldom": "^0.1.27",
+    "canvas": "^2.6.1",
```
Comment (gunzip) [inline package.json:84]: same here

File: src/utils/barcode.ts (hunk 10 above)
Comment (gunzip) [reply src/utils/barcode.ts:25]: or better [taskify](https://github.com/gcanti/fp-ts/blob/1.17.0/src/TaskEither.ts#L325)

File: src/utils/barcode.ts [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+            return resolve(png.toString("base64"));
+          } else {
+            return reject(new Error(`Cannot generate png barcode|${err}`));
+          }
+        })
+      ),
+    errs => new Error(`Cannot generate png barcode|${errs}`)
+  );
+
+const toBase64Svg = (text: string) =>
```
Comment (gunzip) [inline src/utils/barcode.ts:36]: can you try to generate the svg using bwjp as well since it has no external / native deps ?
https://github.com/metafloor/bwip-js/blob/master/examples/svg.html#L150

File: src/utils/barcode.ts [hunk 14]
```diff
@@ -0,0 +1,60 @@
+import * as bwipjs from "bwip-js";
+import { sequenceS } from "fp-ts/lib/Apply";
+import { tryCatch2v } from "fp-ts/lib/Either";
+import { fromEither, taskEither, taskify } from "fp-ts/lib/TaskEither";
+
+export interface IBarcodeOutput {
+  png: string;
+  pngMimeType: string;
```
Comment (gunzip) [inline src/utils/barcode.ts:8]: thre's no need to forward the mime type if the field name is 'png'. the mime type makes sense only if we output a generic list of fields ie.

```ts
attachments: {
  "name": <file name>,
  "data": <content base 64>,
  "mime": <mime type>
}
```

File: src/utils/barcode.ts [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+
+interface IBwipOptions {
+  bcid: string;
+  text: string;
+}
+
+const toBase64Svg = (options: IBwipOptions) =>
+  tryCatch2v(
+    () => {
+      const drawsvg = require("bwip-js/examples/drawing-svg");
```
Comment (gunzip) [inline src/utils/barcode.ts:21]: can we move this require call outside this method at top level ?

File: src/utils/barcode.ts [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+      // tslint:disable-next-line: no-any
+      const anyBwipJs = bwipjs as any;
+      anyBwipJs.fixupOptions(options);
```
Comment (gunzip) [inline src/utils/barcode.ts:24]: can you add some comment and/or link to docs / examples ?

File: src/utils/barcode.ts [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+      return Buffer.from(svg).toString("base64");
+    },
+    errs => new Error(`Cannot generate svg barcode|${errs}`) as Error | string
+  );
+
+const toBase64Png = taskify(bwipjs.toBuffer);
+
+export const toBarcode = (text: string) => {
+  const options = {
+    bcid: "code128",
```
Comment (gunzip) [inline src/utils/barcode.ts:38]: can we make this configurable with an optional environment variable and a default value ?

File: api_backend.yaml [hunk 18]
```diff
@@ -786,6 +789,31 @@ definitions:
...
+    title: MessageAttachment
+    description: Describes a message's attachment
+    properties:
+      name: string
+      content: string
+      mimeType: string
+    required:
+      - name
+      - content
+      - mimeType
```
Comment (AleDore) [inline api_backend.yaml:816]: @Undermaken @gunzip: Can this structure fits our needs?

File: api_backend.yaml [hunk 19]
```diff
@@ -152,18 +152,21 @@ paths:
...
+                id: "01CFAGRMGB9XCA8B2CQ4QA7K76",
+                sender_service_id: "5a25abf4fcc89605c082f042c49a",
+                time_to_live: 3600
               },
-              created_at: "2018-06-06T12:22:24.523Z",
-              fiscal_code: "LSSLCU79B24L219P",
-              id: "01CFAGRMGB9XCA8B2CQ4QA7K76",
-              sender_service_id: "5a25abf4fcc89605c082f042c49a",
-              time_to_live: 3600
+              attachments: [{name:"attachment", content:"aBase64Encoding", mimeType: "image/png"}]
```
Comment (gunzip) [inline api_backend.yaml:169]:
```suggestion
              attachments: [{name:"attachment", content:"aBase64Encoding", mime_type: "image/png"}]
```

File: api_backend.yaml (hunk 18 above)
Comment (gunzip) [inline api_backend.yaml:816]:
```suggestion
      - mime_type
```

File: src/utils/barcode.ts [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+    text
+  };
+
+  return sequenceS(taskEither)({
+    png: toBase64Png(options),
```
[the remark this answers is not in the document]
Comment (AleDore) [reply src/utils/barcode.ts:43]: Yes, cause `bwipjs.toBuffer` returns a `Buffer` simply taskifying with `taskify(bwipjs.toBuffer)`

File: src/utils/barcode.ts (hunk 15 above)
Comment (AleDore) [reply src/utils/barcode.ts:21]: It seems that It could not find a declaration file for this module, using `import * as drawsvg from "bwip-js/examples/drawing-svg";`

File: api_backend.yaml [hunk 21]
```diff
@@ -152,18 +152,21 @@ paths:
...
           description: Found.
           schema:
-            $ref: "#/definitions/CreatedMessageWithContent"
+            $ref: "#/definitions/CreatedMessageWithContentAndAttachments"
           examples:
             application/json: |
-              content: {
-                markdown: "hey hey !! <a style=\"color: red\" href=\"http://example.com\"> some content here ..... this is a link with a style applied, some other content</a>",
-                subject: "my subject ............"
+              message: {
```
Comment (Undermaken) [inline api_backend.yaml:158]: I guess this is a breaking change
Could be **_attachments_** ad optional field added to the previous CreatedMessageWithContent?

Comment (gunzip) [reply api_backend.yaml:158]: of course it's optional: https://github.com/pagopa/io-backend/pull/648/files#diff-aa6cd0613984937586c55eb6ed53d7baR816

File: api_backend.yaml [hunk 22]
```diff
@@ -786,6 +789,32 @@ definitions:
     $ref: "https://raw.githubusercontent.com/pagopa/io-pagopa-proxy/v0.8.6/api_pagopa.yaml#/definitions/SpezzoniCausaleVersamento"
   SpezzoniCausaleVersamentoItem:
     $ref: "https://raw.githubusercontent.com/pagopa/io-pagopa-proxy/v0.8.6/api_pagopa.yaml#/definitions/SpezzoniCausaleVersamentoItem"
+
+  MessageAttachment:
+    type: object
+    title: MessageAttachment
+    description: Describes a message's attachment
+    properties:
+      name: string
```
Comment (Undermaken) [inline api_backend.yaml:798]: should it be in this format?
```suggestion
      name:
        type: string
```

File: api_backend.yaml (hunk 21 above)
Comment (Undermaken) [reply api_backend.yaml:158]: @gunzip 
Maybe I'm missing something

### actual message shape
```json
{
  "content": {
    "markdown": "markdown",
    "subject": "Benvenuto su IO"
  },
  "created_at": "2020-03-24T09:45:44.650Z",
  "fiscal_code": "XXXXTT00A12L719R",
  "id": "01E45YXZYAP49FC7G334ZX7NR8",
  "sender_service_id": "azure-deployc49a",
  "time_to_live": 3600
}
```

### new shape
```json
{
  "message": {
    "content": {
      "markdown": "markdown",
      "subject": "Benvenuto su IO"
    },
    "created_at": "2020-03-24T09:45:44.650Z",
    "fiscal_code": "XXXXTT00A12L719R",
    "id": "01E45YXZYAP49FC7G334ZX7NR8",
    "sender_service_id": "azure-deployc49a",
    "time_to_live": 3600
  },
  "attachments": []
}
```

this PR proposes a wrapping version of the previous message.
Isn't it?

Comment (gunzip) [reply api_backend.yaml:158]: havent noticed that, of course it must not wrap the message

Comment (gunzip) [reply api_backend.yaml:158]: another choice is if we should put attachments into content or aside message metadata

File: api_backend.yaml (hunk 22 above)
Comment (AleDore) [reply api_backend.yaml:798]: Yes I've just correct it this way

File: api_backend.yaml (hunk 21 above)
Comment (AleDore) [reply api_backend.yaml:158]: So we have to modify io-functions-commons and provide a new release to point to new specs

Comment (gunzip) [reply api_backend.yaml:158]: no we don't, those specs are good as is. the message shape here is constrained to the backend-app usage so we just have to modify the message shape here

Comment (AleDore) [reply api_backend.yaml:158]: I don't understand. If we have to shape the message structure based on the last example, we have to modify `MessageContent` spec that has been defined in io-functions-commons. Do we have any mechanism to extend MessageContent in `api_backend.yaml` ?

Comment (AleDore) [reply api_backend.yaml:158]: In this case, providing a new spec for `MessageContent`, i.e: `MessageContentWithAttachments`, which strategy can we adopt to inject this new spec into `CreatedMessageWithContent` ?

Comment (gunzip) [reply api_backend.yaml:158]: we must rewrite it here

File: src/utils/barcode.ts (hunk 20 above)
[the remark this answers is not in the document]
Comment (gunzip) [reply src/utils/barcode.ts:50]: then we should rename it

File: api_backend.yaml (hunk 21 above)
Comment (gunzip) [reply api_backend.yaml:158]: ping:)

File: api_backend.yaml [hunk 23]
```diff
@@ -786,6 +791,52 @@ definitions:
     $ref: "https://raw.githubusercontent.com/pagopa/io-pagopa-proxy/v0.8.6/api_pagopa.yaml#/definitions/SpezzoniCausaleVersamento"
   SpezzoniCausaleVersamentoItem:
     $ref: "https://raw.githubusercontent.com/pagopa/io-pagopa-proxy/v0.8.6/api_pagopa.yaml#/definitions/SpezzoniCausaleVersamentoItem"
+  
+  CreatedMessageWithContentAndAttachments:
+    type: object
+    properties:
```
Comment (gunzip) [inline api_backend.yaml:797]: can you try using allOf with `CreatedMessageWithoutContentAndAttachments` +
`MessageContentWithAttachments` ?

File: src/utils/attachments.ts [hunk 24]
```diff
@@ -0,0 +1,39 @@
+import { flatten } from "fp-ts/lib/Array";
+import { CreatedMessageWithContentAndAttachments } from "generated/backend/CreatedMessageWithContentAndAttachments";
+import { PrescriptionData } from "generated/backend/PrescriptionData";
+import { CreatedMessageWithContent } from "generated/io-api/CreatedMessageWithContent";
+import { ResponseSuccessJson } from "italia-ts-commons/lib/responses";
+import { toBarcode } from "./barcode";
+
+const toMessageAttachments = async (name: string, value: string) => {
```
Comment (gunzip) [inline src/utils/attachments.ts:8]: since this methods are not generic but tied to the barcode generation step, we must use a name (for the file as well) that reflect this link (ie. `toBarcodeAttachments`)

File: src/utils/attachments.ts [hunk 25]
```diff
@@ -0,0 +1,39 @@
...
+  return errorOrBarcodes.fold(
+    () => [],
+    barcodes => [
+      { name, content: barcodes.png, mime_type: barcodes.pngMimeType },
+      { name, content: barcodes.svg, mime_type: barcodes.svgMimeType }
+    ]
+  );
+};
+
+export const fillMessageAttachmentsPayload = async (
```
Comment (gunzip) [inline src/utils/attachments.ts:19]: ditto

Comment (gunzip) [reply src/utils/attachments.ts:19]: moreover this method name does not reflect its implementation: it not only creates the array with the attachments but also wrap its result into a responsesuccessjson (which is not its responsibility; the controller must do that, here it's fair to just return the object)

File: src/utils/attachments.ts [hunk 26]
```diff
@@ -0,0 +1,39 @@
...
+) => {
+  const nreArrays = await toMessageAttachments("nre", prescriptionData.nre);
+  const iupArrays = await toMessageAttachments("iup", prescriptionData.iup);
+  const prescriberArrays = await toMessageAttachments(
+    "prescriber_fiscal_code",
+    prescriptionData.prescriber_fiscal_code as string
+  );
+  const attachments = flatten([nreArrays, iupArrays, prescriberArrays]);
+  return ResponseSuccessJson(
+    CreatedMessageWithContentAndAttachments.encode({
```
Comment (gunzip) [inline src/utils/attachments.ts:31]: just return the object without `ResponseSuccessJson`

File: src/utils/barcode.ts [hunk 27]
```diff
@@ -0,0 +1,69 @@
+import * as bwipjs from "bwip-js";
+import { sequenceS } from "fp-ts/lib/Apply";
+import { Either, left, right, tryCatch2v } from "fp-ts/lib/Either";
+import { fromNullable } from "fp-ts/lib/Option";
+import { fromEither, taskEither, taskify } from "fp-ts/lib/TaskEither";
+import { getRequiredENVVar } from "./container";
+
+const BCID = fromNullable(getRequiredENVVar("BARCODE_ALGO_ID")).getOrElse(
```
Comment (gunzip) [inline src/utils/barcode.ts:8]: if you use `getRequiredENVVar` and `BARCODE_ALGO_ID` is not defined the server won't start and it will throw an error. use `NonEmptyString.decode(process.env.BARCODE_ALGO_ID)` for optional environment variables instead

File: src/utils/barcode.ts (hunk 14 above)
Comment (gunzip) [reply src/utils/barcode.ts:14]: ping (remove mime types from here : )

File: src/utils/barcode.ts [hunk 28]
```diff
@@ -0,0 +1,69 @@
...
+      const svg = anyBwipJs.render(
+        options,
+        drawsvg(options, anyBwipJs.FontLib)
+      );
+      return Buffer.from(svg).toString("base64");
+    },
+    errs => new Error(`Cannot generate svg barcode|${errs}`) as Error | string
+  );
+
+const toBase64Png = taskify(bwipjs.toBuffer);
```
Comment (gunzip) [inline src/utils/barcode.ts:41]: since this method does not return a base64 encoded string, this shouldn't be called `toBase64*`

File: src/utils/barcode.ts (hunk 15 above)
Comment (gunzip) [reply src/utils/barcode.ts:27]: it's ok to just move this outside the method at top level

**Comment by gunzip** [comment]
can you please rebase this on master branch?

File: src/utils/attachments.ts [hunk 29]
```diff
@@ -0,0 +1,40 @@
...
+  return errorOrBarcodes.fold(
+    () => [],
+    barcodes => [
+      { name, content: barcodes.png, mime_type: MIME_TYPES.png },
+      { name, content: barcodes.svg, mime_type: MIME_TYPES.svg }
+    ]
+  );
+};
+
+export const getMessageWithAttachments = async (
```
Comment (gunzip) [inline src/utils/attachments.ts:22]: it's always better to export functions instead of consts, can we convert this into

`export async function getMessageWithAttachments() {... }` ?

File: src/utils/attachments.ts [hunk 30]
```diff
@@ -0,0 +1,40 @@
...
+  message: CreatedMessageWithContent,
+  prescriptionData: PrescriptionData
+) => {
+  const nreArrays = await toBarcodeAttachments("nre", prescriptionData.nre);
```
Comment (gunzip) [inline src/utils/attachments.ts:26]: an optimization would be run these in paralles using `Promise.all` (but the best option is to work with Taks instead of promises and use sequence(task), see https://grossbart.github.io/fp-ts-recipes/#/async?a=work-with-a-list-of-tasks-in-parallel)

to do this 
- return the TaskEither from `toBarcode` (do not call `run`)
- fold the taskEither into a Task and return it from `toBarcodeAttachments`
- sequence Task in getMessageWithAttachments and return the result (do not call run or then, return the array of tasks)
- the controller will call `getMessageWithAttachments(...).map(ResponseSuccessJson).run()`

<-- always try to call `run` at the outermost level and work with Task / TaskEithers internally

File: src/utils/barcode.ts [hunk 31]
```diff
@@ -0,0 +1,66 @@
...
+        drawsvg(options, anyBwipJs.FontLib)
+      );
+      return Buffer.from(svg).toString("base64");
+    },
+    errs => new Error(`Cannot generate svg barcode|${errs}`) as Error | string
+  );
+
+const toBufferPng = taskify(bwipjs.toBuffer);
+
+export const toBarcode = (text: string) => {
```
Comment (gunzip) [inline src/utils/barcode.ts:42]: use `export function` instead and type the return value

File: src/utils/attachments.ts (hunk 30 above)
Comment (AleDore) [reply src/utils/attachments.ts:26]: Actually by removing `run` from `toBarcode` it returns `Task<Either<Error, IBarcodeOutput>>`

Comment (AleDore) [reply src/utils/attachments.ts:26]: Edit: Nevermind I've refactored the code by change `fold` to `foldTaskEither` in `toBarcode` function

File: src/utils/attachments.ts [hunk 32]
```diff
@@ -0,0 +1,37 @@
...
+import { MessageAttachment } from "generated/backend/MessageAttachment";
+import { PrescriptionData } from "generated/backend/PrescriptionData";
+import { toBarcode } from "./barcode";
+
+const MIME_TYPES = {
+  png: "image/png",
+  svg: "image/svg+xml"
+};
+
+const toBarcodeAttachments = (name: string, value: string) => {
```
Comment (gunzip) [inline src/utils/attachments.ts:12]: can you add a comment to this method to explain what `name` and `value` represents ?

Comment (gunzip) [reply src/utils/attachments.ts:12]: what's the inferred return type of this method ? (it should be `MessageAttachment[]`, if not you must specify it)

File: src/utils/attachments.ts [hunk 33]
```diff
@@ -0,0 +1,37 @@
...
+  return errorOrBarcodes.fold(
+    () => [],
+    barcodes => [
+      { name, content: barcodes.png, mime_type: MIME_TYPES.png },
+      { name, content: barcodes.svg, mime_type: MIME_TYPES.svg }
+    ]
+  );
+};
+
+export function getAttachments(
```
Comment (gunzip) [inline src/utils/attachments.ts:23]: since this is not a generic method, I'd call it something like `renderPrescriptionAttachments` or `getPrescriptionAttachments`

File: src/utils/attachments.ts [hunk 34]
```diff
@@ -0,0 +1,37 @@
...
+  prescriptionData: PrescriptionData
+  // tslint:disable-next-line: readonly-array
+): Task<MessageAttachment[]> {
+  return array
+    .sequence(task)([
+      toBarcodeAttachments("iup", prescriptionData.iup),
```
Comment (gunzip) [inline src/utils/attachments.ts:29]: I'm not sure but maybe we can use `PrescriptionData.props` here instead of hardconding the parameter names, can you check it out ?

File: src/utils/barcode.ts [hunk 35]
```diff
@@ -0,0 +1,73 @@
...
+import { left, right, tryCatch2v } from "fp-ts/lib/Either";
+import {
+  fromEither,
+  taskEither,
+  TaskEither,
+  taskify
+} from "fp-ts/lib/TaskEither";
+import { NonEmptyString } from "italia-ts-commons/lib/strings";
+
+const BCID = NonEmptyString.decode(process.env.BARCODE_ALGO_ID).getOrElse(
```
Comment (gunzip) [inline src/utils/barcode.ts:12]: `process.env` should be accessed only by `config.ts` (otherwise that would be hard to track this global configuration values)
so this value must be moved there (I'd give it a more explicit name ie'. `BARCODE_ALGORITHM`

File: src/utils/barcode.ts [hunk 36]
```diff
@@ -0,0 +1,73 @@
...
+          typeof errorOrString === "string"
+            ? new Error(errorOrString)
+            : errorOrString
+        )
+      ),
+    images =>
+      fromEither(
+        right({
+          png: images.png.toString("base64"),
+          svg: images.svg
```
Comment (gunzip) [inline src/utils/barcode.ts:69]: that would be better to move `svg.toString("base64")` here for simmetry and use `toBufferSvg` instead of
 `toBase64Svg`

File: src/utils/attachments.ts (hunk 34 above)
Comment (AleDore) [reply src/utils/attachments.ts:29]: It seems that props does not exists on `PrescriptionData`

File: src/services/__tests__/messagesService.test.ts [hunk 37]
```diff
@@ -244,6 +283,24 @@ describe("MessageService#getMessage", () => {
...
   });
 
+  it("returns a message with attachments from the API", async () => {
+    mockGetMessage.mockImplementation(() =>
+      t.success(validApiWithAttachmentsMessageResponse)
+    );
+
+    const service = new MessageService(api);
+
+    const res = await service.getMessage(mockedUser, aValidMessageId);
```
Comment (gunzip) [inline src/services/__tests__/messagesService.test.ts:293]: this test should mock `messageModel.getMessage()` to return a message with some
 `prescription_data` and NOT mock `service.getMessage()` (otherwise we're testing nothing here)

File: src/services/__tests__/messagesService.test.ts [hunk 38]
```diff
@@ -68,6 +82,29 @@ const validApiMessageResponse = {
     status: "PROCESSED"
   }
 };
+
+const validApiWithAttachmentsMessageResponse = {
+  status: 200,
+  value: {
+    message: {
+      content: {
+        attachments: [aValidNrePngBarcode, aValidNreSvgBarcode],
```
Comment (gunzip) [inline src/services/__tests__/messagesService.test.ts:91]: sorry I've review the test and my first suggestion was wrong. the test must be rewritten anyway:
`validApiWithAttachmentsMessageResponse` is the response we get from io-functions-app which does not contains the `attachments` field BUT contains the `prescription_metadata` values. the test must state that when we get `prescription_metadata` in input we get the correct `attachments` output

Comment (gunzip) [reply src/services/__tests__/messagesService.test.ts]: fixed here: https://github.com/pagopa/io-backend/pull/648/commits/c18e7cb45346d7c8b7d11f09a99b389c371bb50b?file-filters%5B%5D=.ts

File: src/services/__tests__/messagesService.test.ts [hunk 39]
```diff
@@ -68,6 +69,33 @@ const validApiMessageResponse = {
...
 };
+
+const validApiMessageResponseWithPrescriptionMetadata = {
+  status: 200,
+  value: {
+    message: {
+      content: {
+        markdown: aValidMarkdown,
+        prescription_data: {
+          ipe: "12345678",
```
Comment (Undermaken) [inline src/services/__tests__/messagesService.test.ts:80]:
```suggestion
          iup: "12345678",
```

File: src/utils/__tests__/barcode.test.ts [hunk 40]
```diff
@@ -0,0 +1,17 @@
...
+import { toBarcode } from "../barcode";
+const aText = "FRLFRC74E04B157I";
+
+describe("Barcode generator", () => {
+  it("should return a valid svg and png", async () => {
+    const errorOrBarcodes = await toBarcode(aText).run();
+    expect(isRight(errorOrBarcodes)).toBeTruthy();
+    expect(errorOrBarcodes.value).toMatchSnapshot();
+  });
+});
```
Comment (Undermaken) [inline src/utils/__tests__/barcode.test.ts:17]: Could be nice to add a test with empty string (spoiler: it fails, the test passes)
```typescript
it("should fail with empty string", async () => {
    const errorOrBarcodes = await toBarcode("").run();
    expect(isLeft(errorOrBarcodes)).toBeTruthy();
  });
```
